leaderelection/context.go

```diff
@@ -0,0 +1,225 @@
+
+type standardBuilder struct {
+	kc  kubernetes.Interface
+	lec ComponentConfig
+}
+
+func (b *standardBuilder) BuildElector(ctx context.Context, la reconciler.LeaderAware, name string, enq func(reconciler.Bucket, types.NamespacedName)) (Elector, error) {
+	logger := logging.FromContext(ctx)
+
+	id, err := UniqueID()
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := make([]Elector, 0, b.lec.Buckets)
+	for i := uint32(0); i < b.lec.Buckets; i++ {
+		bkt := &bucket{
+			component: b.lec.Component,
+			name:      name,
+			index:     i,
+			total:     b.lec.Buckets,
+		}
+
+		rl, err := resourcelock.New(b.lec.ResourceLock,
+			system.Namespace(), // use namespace we are running in
+			bkt.Name(),
+			b.kc.CoreV1(),
+			b.kc.CoordinationV1(),
+			resourcelock.ResourceLockConfig{
+				Identity: id,
+			})
+		if err != nil {
+			return nil, err
+		}
+		logger.Infof("%s will run in leader-elected mode with id %q", bkt.Name(), rl.Identity())
+
+		le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: b.lec.LeaseDuration,
+			RenewDeadline: b.lec.RenewDeadline,
+			RetryPeriod:   b.lec.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(context.Context) {
+					logger.Infof("%q has started leading %q", rl.Identity(), bkt.Name())
+					la.Promote(bkt, enq)
+				},
+				OnStoppedLeading: func() {
+					logger.Infof("%q has stopped leading %q", rl.Identity(), bkt.Name())
+					la.Demote(bkt)
+				},
+			},
+			ReleaseOnCancel: true,
+
+			// TODO: use health check watchdog, knative/pkg#1048
+			Name: b.lec.Component,
+		})
+		if err != nil {
+			return nil, err
+		}
+		// TODO: use health check watchdog, knative/pkg#1048
+		// if lec.WatchDog != nil {
+		// 	lec.WatchDog.SetLeaderElection(le)
+		// }
+		buckets = append(buckets, &runUntilCancelled{le: le})
+	}
+	return &runAll{les: buckets}, nil
+}
+
+// unopposedElector promotes when run without needing to be elected.
+type unopposedElector struct {
+	bkt reconciler.Bucket
+	la  reconciler.LeaderAware
+	enq func(reconciler.Bucket, types.NamespacedName)
+}
+
+// Run implements Elector
+func (ue *unopposedElector) Run(ctx context.Context) {
+	ue.la.Promote(ue.bkt, ue.enq)
+}
+
+type runAll struct {
+	les []Elector
+}
+
+// Run implements Elector
+func (ra *runAll) Run(ctx context.Context) {
+	sg := sync.WaitGroup{}
+	defer sg.Wait()
+
+	for _, le := range ra.les {
+		sg.Add(1)
+		go func(le Elector) {
+			defer sg.Done()
+			le.Run(ctx)
+		}(le)
+	}
+}
+
+// runUntilCancelled wraps a single-term Elector into one that runs until
+// the passed context is cancelled.
+type runUntilCancelled struct {
+	le Elector
+}
+
+// Run implements Elector
+func (ruc *runUntilCancelled) Run(ctx context.Context) {
+	// Run executes a single iteration of an election:
+	// 1. Wait to become leader
+	// 2. Call OnStartedLeading
+	// 3. Wait for ctx cancel or loss of leader.
+	// 4. Call OnStoppedLeading
```

Comment:
This is only correct when the field `le` is a LeadElector of k8s.io/client-go/tools/leaderelect. Maybe make it clear?